crates/store/src/state/state_db.rs

```diff
@@ -1,60 +1,267 @@
-    transaction::StoreTransaction,
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Mutex,
 };
-use anyhow::{anyhow, Result};
-use gw_common::{error::Error as StateError, smt::SMT, state::State, H256};
-use gw_db::schema::{COLUMN_DATA, COLUMN_SCRIPT};
+
+use anyhow::Result;
 use gw_traits::CodeStore;
 use gw_types::{
     bytes::Bytes,
-    from_box_should_be_ok,
-    packed::{self, AccountMerkleState, Byte32},
-    prelude::*,
+    packed::{self, LogItem},
+    prelude::{Pack, Unpack},
 };
-use log::log_enabled;
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
-pub enum StateContext {
-    ReadOnly,
-    ReadOnlyHistory(u64),
-    AttachBlock(u64),
-    DetachBlock(u64),
+
+use gw_common::{error::Error as StateError, smt::SMT, state::State, H256};
+
+use crate::{
+    smt::smt_store::SMTStateStore,
+    snapshot::StoreSnapshot,
+    state::history::history_state::{HistoryState, HistoryStateStore},
+    traits::{chain_store::ChainStore, kv_store::KVStore},
+};
+
+use super::{
+    history::history_state::RWConfig,
+    overlay::{mem_state::MemStateTree, mem_store::MemStore},
+    traits::JournalDB,
+    BlockStateDB, MemStateDB,
+};
+
+#[derive(Debug, Clone)]
+pub enum JournalEntry {
+    UpdateRaw { key: H256, prev_value: H256 },
+    SetAccountCount { prev_count: u32 },
+    InsertScript { script_hash: H256, prev_exist: bool },
+    InsertData { data_hash: H256, prev_exist: bool },
+    AppendLog { index: usize },
+}
+
+impl JournalEntry {
+    /// revert journal
+    pub fn revert<S>(self, state_db: &mut StateDB<S>) {
+        use JournalEntry::*;
+
+        match self {
+            UpdateRaw { key, prev_value } => {
+                state_db.dirty_state.insert(key, prev_value);
+            }
+            SetAccountCount { prev_count } => {
+                state_db.dirty_account_count = Some(prev_count);
+            }
+            InsertScript {
+                script_hash,
+                prev_exist,
+            } => {
+                if !prev_exist {
+                    state_db.dirty_scripts.remove(&script_hash);
+                }
+            }
+            InsertData {
+                data_hash,
+                prev_exist,
+            } => {
+                if !prev_exist {
+                    state_db.dirty_data.remove(&data_hash);
+                }
+            }
+            AppendLog { index } => {
+                assert_eq!(state_db.dirty_logs.len(), index + 1);
+                state_db.dirty_logs.pop();
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct Journal {
+    entries: Vec<JournalEntry>,
+}
+
+impl Journal {
+    pub fn len(&self) -> usize {
+        self.entries.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.entries.is_empty()
+    }
+
+    /// Revert journal to len
+    /// Return reverted entries sorted in journal emit order
+    pub fn revert_entries(&mut self, len: usize) -> Vec<JournalEntry> {
+        self.entries.split_off(len)
```

Comment:
Is it possible to use `drain(len..).rev()` instead? That doesn't need to allocate and copy all the reverted entries.

Reply:
I tried it and the borrow checker is not happy. But it can be resolved if the `JournalEntry::revert` method is inlined.